// @astrojs/ts-plugin/dist/project-astro-files.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var project_astro_files_exports = {};
__export(project_astro_files_exports, {
  ProjectAstroFilesManager: () => ProjectAstroFilesManager
});
module.exports = __toCommonJS(project_astro_files_exports);
var import_utils = require("./utils");
const _ProjectAstroFilesManager = class {
  constructor(typescript, project, serverHost, snapshotManager, parsedCommandLine) {
    this.typescript = typescript;
    this.project = project;
    this.serverHost = serverHost;
    this.snapshotManager = snapshotManager;
    this.parsedCommandLine = parsedCommandLine;
    this.files = /* @__PURE__ */ new Set();
    this.directoryWatchers = /* @__PURE__ */ new Set();
    this.setupWatchers();
    this.updateProjectAstroFiles();
    _ProjectAstroFilesManager.instances.set(project.getProjectName(), this);
  }
  static getInstance(projectName) {
    return this.instances.get(projectName);
  }
  updateProjectConfig(serviceHost) {
    var _a;
    const parsedCommandLine = (_a = serviceHost.getParsedCommandLine) == null ? void 0 : _a.call(serviceHost, (0, import_utils.getConfigPathForProject)(this.project));
    if (!parsedCommandLine) {
      return;
    }
    this.disposeWatchersAndFiles();
    this.parsedCommandLine = parsedCommandLine;
    this.setupWatchers();
    this.updateProjectAstroFiles();
  }
  getFiles() {
    return Array.from(this.files);
  }
  setupWatchers() {
    for (const directory in this.parsedCommandLine.wildcardDirectories) {
      if (!Object.prototype.hasOwnProperty.call(this.parsedCommandLine.wildcardDirectories, directory)) {
        continue;
      }
      const watchDirectoryFlags = this.parsedCommandLine.wildcardDirectories[directory];
      const watcher = this.serverHost.watchDirectory(
        directory,
        this.watcherCallback.bind(this),
        watchDirectoryFlags === this.typescript.WatchDirectoryFlags.Recursive,
        this.parsedCommandLine.watchOptions
      );
      this.directoryWatchers.add(watcher);
    }
  }
  watcherCallback(fileName) {
    if (!(0, import_utils.isAstroFilePath)(fileName)) {
      return;
    }
    this.updateProjectAstroFiles();
  }
  updateProjectAstroFiles() {
    const fileNamesAfter = (0, import_utils.readProjectAstroFilesFromFs)(this.typescript, this.project, this.parsedCommandLine);
    const removedFiles = new Set(...this.files);
    const newFiles = fileNamesAfter.filter((fileName) => {
      const has = this.files.has(fileName);
      if (has) {
        removedFiles.delete(fileName);
      }
      return !has;
    });
    for (const newFile of newFiles) {
      this.addFileToProject(newFile);
      this.files.add(newFile);
    }
    for (const removedFile of removedFiles) {
      this.removeFileFromProject(removedFile, false);
      this.files.delete(removedFile);
    }
  }
  addFileToProject(newFile) {
    this.snapshotManager.create(newFile);
    const snapshot = this.project.projectService.getScriptInfo(newFile);
    if (snapshot) {
      this.project.addRoot(snapshot);
    }
  }
  removeFileFromProject(file, exists = true) {
    const info = this.project.getScriptInfo(file);
    if (info) {
      this.project.removeFile(info, exists, true);
    }
  }
  disposeWatchersAndFiles() {
    this.directoryWatchers.forEach((watcher) => watcher.close());
    this.directoryWatchers.clear();
    this.files.forEach((file) => this.removeFileFromProject(file));
    this.files.clear();
  }
  dispose() {
    this.disposeWatchersAndFiles();
    _ProjectAstroFilesManager.instances.delete(this.project.getProjectName());
  }
};
let ProjectAstroFilesManager = _ProjectAstroFilesManager;
ProjectAstroFilesManager.instances = /* @__PURE__ */ new Map();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ProjectAstroFilesManager
});
