// @astrojs/ts-plugin/dist/utils.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  ensureRealAstroFilePath: () => ensureRealAstroFilePath,
  getConfigPathForProject: () => getConfigPathForProject,
  isAstroFilePath: () => isAstroFilePath,
  isNotNullOrUndefined: () => isNotNullOrUndefined,
  isVirtualAstroFilePath: () => isVirtualAstroFilePath,
  readProjectAstroFilesFromFs: () => readProjectAstroFilesFromFs,
  replaceDeep: () => replaceDeep,
  toRealAstroFilePath: () => toRealAstroFilePath
});
module.exports = __toCommonJS(utils_exports);
function isAstroFilePath(filePath) {
  return filePath.endsWith(".astro");
}
function isVirtualAstroFilePath(filePath) {
  return filePath.endsWith(".astro.ts");
}
function toRealAstroFilePath(filePath) {
  return filePath.slice(0, -".ts".length);
}
function ensureRealAstroFilePath(filePath) {
  return isVirtualAstroFilePath(filePath) ? toRealAstroFilePath(filePath) : filePath;
}
function isNotNullOrUndefined(val) {
  return val !== void 0 && val !== null;
}
function replaceDeep(obj, searchStr, replacementStr) {
  return _replaceDeep(obj);
  function _replaceDeep(_obj) {
    if (typeof _obj === "string") {
      return _obj.replace(searchStr, replacementStr);
    }
    if (Array.isArray(_obj)) {
      return _obj.map((entry) => _replaceDeep(entry));
    }
    if (typeof _obj === "object") {
      return Object.keys(_obj).reduce((_o, key) => {
        _o[key] = _replaceDeep(_obj[key]);
        return _o;
      }, {});
    }
    return _obj;
  }
}
function getConfigPathForProject(project) {
  var _a;
  return (_a = project.canonicalConfigFilePath) != null ? _a : project.getCompilerOptions().configFilePath;
}
function readProjectAstroFilesFromFs(ts, project, parsedCommandLine) {
  const fileSpec = parsedCommandLine.raw;
  const { include, exclude } = fileSpec;
  if ((include == null ? void 0 : include.length) === 0) {
    return [];
  }
  return ts.sys.readDirectory(project.getCurrentDirectory() || process.cwd(), [".astro"], exclude, include).map(ts.server.toNormalizedPath);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ensureRealAstroFilePath,
  getConfigPathForProject,
  isAstroFilePath,
  isNotNullOrUndefined,
  isVirtualAstroFilePath,
  readProjectAstroFilesFromFs,
  replaceDeep,
  toRealAstroFilePath
});
